Improve Tab Map performance

Anthony-Eid opened the pull request:
## Context

While looking into: #32051 and #16120 with instruments, I noticed that `TabSnapshot::to_tab_point` and `TabSnapshot::to_fold_point` are a common bottleneck between the two issues. This PR takes the first steps into closing the stated issues by improving the performance of both those functions.

### Method

`to_tab_point` and `to_fold_point` iterate through each character in their rows to find tab characters and translate those characters into their respective transformations. This PR changes this iteration to take advantage of the tab character bitmap in the `Rope` data structure and goes directly to each tab character when iterating. 

The tab bitmap is now passed from each layer in-between the `Rope` to the `TabMap`. 

### Testing 

I added several randomized tests to ensure that the new `to_tab_point` and `to_fold_point` functions have the same behavior as the old methods they're replacing. I also added `test_random_chunk_bitmap` on each layer the tab bitmap is passed up to the `TabMap` to make sure that the bitmap being passed is transformed correctly between the layers of `DisplayMap`.

`test_random_chunk_bitmap` was added to these layers:
- buffer
- multi buffer
- custom_highlights
- inlay_map
- fold_map

## Benchmarking 

I setup benchmarks with criterion that is runnable via `cargo bench -p editor  --profile=release-fast`. When benchmarking I had my laptop plugged in and did so from the terminal with a minimal amount of processes running. I'm also on a m4 max

### Results 

#### To Tab Point

Went from completing 6.8M iterations in 5s with an average time of `736.13 ns` to `683.38 ns` which is a `-7.1875%` improvement

#### To Fold Point

Went from completing 6.8M iterations in 5s with an average time of `736.55 ns` to `682.40 ns` which is a `-7.1659%` improvement

#### Editor render 

Went from having an average render time of `62.561  µs` to `57.216 µs` which is a `-8.8248%` improvement

#### Build Buffer with one long line

Went from having an average buffer build time of `3.2549 ms` to `3.2635 ms` which is a `+0.2151%` regression within the margin of error

#### Editor with 1000 multi cursor input 

Went from having an average edit time of `133.05 ms` to `122.96 ms` which is a `-7.5776%` improvement

Release Notes:

- N/A

Anthony-Eid committed:
Start work on optomizing tab map

We currently iterate over each character when looking for tab bytes even
though chunks keeps a bitmask that represents each tab position. This
commit is the first step in using the bitmask

Co-authored-by: Remco Smits <djsmits12@gmail.com>
Co-authored-by: Cole Miller <m@cole-miller.net>

Anthony-Eid committed:
more wip

Anthony-Eid committed:
add tests

Anthony-Eid committed:
Get every test passing except random tabs

Anthony-Eid committed:
fix repro panic

Anthony-Eid committed:
Simplify tab cursor code

Anthony-Eid committed:
Get tab cursor working with correct character offset with utf16 chars

Anthony-Eid committed:
Get all test passing

Anthony-Eid committed:
Implement expand tabs (will follow up with tests)

Anthony-Eid committed:
Merge remote-tracking branch 'origin/main' into tab-map-bitmask

Anthony-Eid committed:
Bug fix of incorrect tabs/chars bitmap shift

Anthony-Eid committed:
Add testing suite for chunk bitmaps

Anthony-Eid committed:
Add multi buffer test for tab map work

Anthony-Eid committed:
Start work on bench marking

Co-authored-by: Cole Miller <cole@zed.dev>

Anthony-Eid committed:
Switch to view

Anthony-Eid committed:
WIP and merge

Anthony-Eid committed:
More wip

Anthony-Eid committed:
Merge remote-tracking branch 'origin' into tab-map-bitmask

Anthony-Eid committed:
Get inlay iterator to pass bitmaps correctly

Anthony-Eid committed:
Get highlight map to pass

Anthony-Eid committed:
merge

Anthony-Eid committed:
improve expand tabs counting char byte overshoot

Anthony-Eid committed:
Fix clippy errors

Anthony-Eid committed:
Fix to_fold_point bug

The bug was due to TabStopCursor incorrectly counting char's offset in
an edge case. Which has been fixed

Anthony-Eid committed:
Improve bench mark suite

Anthony-Eid committed:
Merge branch 'main' into tab-map-bitmask

Anthony-Eid committed:
Add editor bench mark for multi cursor edits

Anthony-Eid committed:
Merge remote-tracking branch 'origin/main' into tab-map-bitmask

Anthony-Eid committed:
Introduce ChunkBitmaps struct for text, chars and tabs

The change introduces a dedicated struct to hold chunk bitmaps data
rather than using raw tuples.

Anthony-Eid commented on crates/rope/src/rope.rs:
```diff
@@ -821,6 +873,30 @@ impl<'a> Chunks<'a> {
+
+impl<'a> Iterator for ChunkWithBitmaps<'a> {
+    /// text, chars bitmap, tabs bitmap
+    type Item = (&'a str, u128, u128);
```
Fixed with `ChunkBitmaps`

Anthony-Eid committed:
Start adding bench marks for to_tab_point

Co-authored-by: Piotr Osiewicz <24362066+osiewicz@users.noreply.github.com>

SomeoneToIgnore reviewed (commented):
Added a few notes aside to the main task.

Anthony-Eid committed:
Merge remote-tracking branch 'origin/main' into tab-map-bitmask

Anthony-Eid committed:
Use origin main's languges/src/typescript file

Anthony-Eid committed:
Fix to_tab_point benchmark suite

Anthony-Eid committed:
Improve TabStopCursor performance by making the iterator a generic

The internal iterator of TabStopCursor was a Box<dyn Iterator>, this
caused a good amount of overhead and prevented the compiler from
inlining the iterator. Before this change the benchmark could only do
5.5M iterations, now it does 7.1M iterations.

The implementation this replaces does 6.8M iterations

Anthony-Eid committed:
Add a benchmark for to fold point

Anthony-Eid committed:
Update 1000 multicursor bench

Anthony-Eid commented on crates/languages/src/typescript.rs:
```diff
@@ -780,8 +780,8 @@ pub struct EsLintLspAdapter {
 impl EsLintLspAdapter {
-    const CURRENT_VERSION: &'static str = "2.4.4";
-    const CURRENT_VERSION_TAG_NAME: &'static str = "release/2.4.4";
+    const CURRENT_VERSION: &'static str = "3.0.10";
```
fixed

Anthony-Eid commented on crates/language_tools/src/lsp_log.rs:
```diff
@@ -436,7 +436,7 @@ impl LogStore {
             id,
             LogMessage {
-                message: message.trim_end().to_string(),
+                message: message.trim().to_string(),
```
fixed

Anthony-Eid committed:
Fix clippy errors

Anthony-Eid committed:
Remove todos and fixmes

I left a comment on the PR for the todo i had a question for. The rest of them weren't relevent anymore
